[Bug]: aws_elasticache_user_group_association fails on concurrent updates

### Terraform Core Version

1.3.7

### AWS Provider Version

4.48.0

### Affected Resource(s)

* aws_elasticache_user_group_association

### Expected Behavior

It should be possible to apply two modules containing `aws_elasticache_user_group_association` resources concurrently.

### Actual Behavior

The apply in one module fails with the error "InvalidUserGroupState: User group test-group has status modifying. Only active user groups can be modified."

### Relevant Error/Panic Output Snippet

_No response_

### Terraform Configuration Files

A base module that creates a group and two users:

```
provider "aws" {
  region = "eu-west-1"
}

resource "aws_elasticache_user" "one" {
  user_id       = "one"
  user_name     = "one"
  access_string = "on -@all"
  engine        = "REDIS"
  no_password_required = true
}

resource "aws_elasticache_user" "two" {
  user_id       = "two"
  user_name     = "two"
  access_string = "on -@all"
  engine        = "REDIS"
  no_password_required = true
}

resource "aws_elasticache_user_group" "this" {
  engine        = "REDIS"
  user_group_id = "test-group"

  user_ids = ["default"]

  lifecycle {
    ignore_changes = ["user_ids"]
  }
}
```

A module that creates an association:

```
provider "aws" {
  region = "eu-west-1"
}

resource "aws_elasticache_user_group_association" "one" {
  user_group_id = "test-group"
  user_id = "one"
}
```

Another module that creates an association:

```
provider "aws" {
  region = "eu-west-1"
}

resource "aws_elasticache_user_group_association" "two" {
  user_group_id = "test-group"
  user_id = "two"
}
```


### Steps to Reproduce

First, apply the base module. Then apply the two later modules concurrently. This can be done in the shell with a script like this:

```
#!/bin/bash

set -e

for i in user-1 user-2; do
    terraform -chdir="$i" apply -auto-approve &
done

wait
```

The apply is slow enough that it could also be done using different terminal windows.

### Debug Output

```
aws_elasticache_user_group_association.two: Creating...
2023-01-05T11:08:45.070Z [INFO]  Starting apply for aws_elasticache_user_group_association.two
2023-01-05T11:08:45.070Z [DEBUG] aws_elasticache_user_group_association.two: applying the planned Create change
2023-01-05T11:08:45.070Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: [DEBUG] Waiting for state to become: [success]
2023-01-05T11:08:45.070Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: [DEBUG] [aws-sdk-go] DEBUG: Request elasticache/ModifyUserGroup Details:
2023-01-05T11:08:45.070Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: ---[ REQUEST POST-SIGN ]-----------------------------
2023-01-05T11:08:45.070Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: POST / HTTP/1.1
2023-01-05T11:08:45.070Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: Host: elasticache.eu-west-1.amazonaws.com
2023-01-05T11:08:45.070Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: User-Agent: APN/1.0 HashiCorp/1.0 Terraform/1.3.7 (+https://www.terraform.io) terraform-provider-aws/4.48.0 (+https://registry.terraform.io/providers/hashicorp/aws) aws-sdk-go/1.44.162 (go1.19.3; linux; amd64)
2023-01-05T11:08:45.070Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: Content-Length: 90
2023-01-05T11:08:45.070Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: Authorization: xxx
2023-01-05T11:08:45.070Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: Content-Type: application/x-www-form-urlencoded; charset=utf-8
2023-01-05T11:08:45.070Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: X-Amz-Date: 20230105T110845Z
2023-01-05T11:08:45.070Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: X-Amz-Security-Token: xxx
2023-01-05T11:08:45.070Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: Accept-Encoding: gzip
2023-01-05T11:08:45.070Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: 
2023-01-05T11:08:45.070Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: Action=ModifyUserGroup&UserGroupId=test-group&UserIdsToAdd.member.1=two&Version=2015-02-02
2023-01-05T11:08:45.070Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: -----------------------------------------------------
2023-01-05T11:08:45.356Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: [DEBUG] [aws-sdk-go] DEBUG: Response elasticache/ModifyUserGroup Details:
2023-01-05T11:08:45.356Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: ---[ RESPONSE ]--------------------------------------
2023-01-05T11:08:45.356Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: HTTP/1.1 200 OK
2023-01-05T11:08:45.356Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: Content-Length: 702
2023-01-05T11:08:45.356Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: Content-Type: text/xml
2023-01-05T11:08:45.356Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: Date: Thu, 05 Jan 2023 11:08:44 GMT
2023-01-05T11:08:45.356Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: X-Amzn-Requestid: e8f1dba2-3d9f-4fac-bf4b-5402f737d078
2023-01-05T11:08:45.356Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: 
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: 
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: -----------------------------------------------------
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: [DEBUG] [aws-sdk-go] <ModifyUserGroupResponse xmlns="http://elasticache.amazonaws.com/doc/2015-02-02/">
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:   <ModifyUserGroupResult>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:     <UserGroupId>test-group</UserGroupId>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:     <PendingChanges>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:       <UserIdsToAdd>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:         <member>one</member>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:       </UserIdsToAdd>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:     </PendingChanges>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:     <MinimumEngineVersion>6.0</MinimumEngineVersion>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:     <ReplicationGroups/>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:     <Engine>redis</Engine>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:     <UserIds>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:       <member>default</member>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:     </UserIds>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:     <ARN>arn:aws:elasticache:eu-west-1:000000000000:usergroup:test-group</ARN>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:     <Status>modifying</Status>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:   </ModifyUserGroupResult>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:   <ResponseMetadata>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:     <RequestId>e8f1dba2-3d9f-4fac-bf4b-5402f737d078</RequestId>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:   </ResponseMetadata>
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: </ModifyUserGroupResponse>
2023-01-05T11:08:45.357Z [INFO]  provider.terraform-provider-aws_v4.48.0_x5: [INFO] Waiting for ElastiCache User Group (test-group,one) to be available
2023-01-05T11:08:45.357Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: [DEBUG] Waiting for state to become: [active]
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: [DEBUG] [aws-sdk-go] DEBUG: Response elasticache/ModifyUserGroup Details:
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: ---[ RESPONSE ]--------------------------------------
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: HTTP/1.1 400 Bad Request
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: Connection: close
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: Content-Length: 344
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: Content-Type: text/xml
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: Date: Thu, 05 Jan 2023 11:08:44 GMT
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: X-Amzn-Requestid: cc8ba6b3-929b-424f-b745-7e3cd6318454
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: 
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: 
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: -----------------------------------------------------
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: [DEBUG] [aws-sdk-go] <ErrorResponse xmlns="http://elasticache.amazonaws.com/doc/2015-02-02/">
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:   <Error>
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:     <Type>Sender</Type>
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:     <Code>InvalidUserGroupState</Code>
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:     <Message>User group test-group has status modifying. Only active user groups can be modified.</Message>
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:   </Error>
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:   <RequestId>cc8ba6b3-929b-424f-b745-7e3cd6318454</RequestId>
2023-01-05T11:08:45.389Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: </ErrorResponse>
2023-01-05T11:08:45.390Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5: [DEBUG] [aws-sdk-go] DEBUG: Validate Response elasticache/ModifyUserGroup failed, attempt 0/25, error InvalidUserGroupState: User group test-group has status modifying. Only active user groups can be modified.
2023-01-05T11:08:45.390Z [DEBUG] provider.terraform-provider-aws_v4.48.0_x5:    status code: 400, request id: cc8ba6b3-929b-424f-b745-7e3cd6318454
2023-01-05T11:08:45.390Z [ERROR] provider.terraform-provider-aws_v4.48.0_x5: Response contains error diagnostic: tf_rpc=ApplyResourceChange diagnostic_summary="creating ElastiCache User Group Association ("test-group,two"): InvalidUserGroupState: User group test-group has status modifying. Only active user groups can be modified.
        status code: 400, request id: cc8ba6b3-929b-424f-b745-7e3cd6318454" tf_proto_version=5.3 tf_provider_addr=registry.terraform.io/hashicorp/aws tf_req_id=8ea6a3ef-36de-f65d-7c3b-8715b37e4374 @caller=github.com/hashicorp/terraform-plugin-go@v0.14.2/tfprotov5/internal/diag/diagnostics.go:55 @module=sdk.proto diagnostic_detail= diagnostic_severity=ERROR tf_resource_type=aws_elasticache_user_group_association timestamp=2023-01-05T11:08:45.389Z
2023-01-05T11:08:45.402Z [ERROR] vertex "aws_elasticache_user_group_association.two" error: creating ElastiCache User Group Association ("test-group,two"): InvalidUserGroupState: User group test-group has status modifying. Only active user groups can be modified.
        status code: 400, request id: cc8ba6b3-929b-424f-b745-7e3cd6318454
╷
│ Error: creating ElastiCache User Group Association ("test-group,two"): InvalidUserGroupState: User group test-group has status modifying. Only active user groups can be modified.
│       status code: 400, request id: cc8ba6b3-929b-424f-b745-7e3cd6318454
│ 
│   with aws_elasticache_user_group_association.two,
│   on main.tf line 5, in resource "aws_elasticache_user_group_association" "two":
│    5: resource "aws_elasticache_user_group_association" "two" {
│ 
╵
```

### Panic Output

_No response_

### Important Factoids

This gets triggered on a development environment where developers can create users for their own testing. If two developers try to create users at the same time, one or the other fails.

This is made worse because it seems to take several minutes to add users to a group.

### References

_No response_

### Would you like to implement a fix?

Yes

--- a/.changelog/28689.txt
+++ b/.changelog/28689.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_elasticache_user_group_association: Retry on `InvalidUserGroupState` errors to handle concurrent updates
+```

--- a/internal/service/elasticache/user_group_association.go
+++ b/internal/service/elasticache/user_group_association.go
@@ -49,9 +49,11 @@ func resourceUserGroupAssociationCreate(d *schema.ResourceData, meta interface{}
 
 	id := userGroupAssociationID(d.Get("user_group_id").(string), d.Get("user_id").(string))
 
-	_, err := tfresource.RetryWhenNotFound(30*time.Second, func() (interface{}, error) {
-		return conn.ModifyUserGroup(input)
-	})
+	_, err := tfresource.RetryWhenAWSErrCodeEquals(10*time.Minute, func() (interface{}, error) {
+		return tfresource.RetryWhenNotFound(30*time.Second, func() (interface{}, error) {
+			return conn.ModifyUserGroup(input)
+		})
+	}, elasticache.ErrCodeInvalidUserGroupStateFault)
 
 	if err != nil {
 		return fmt.Errorf("creating ElastiCache User Group Association (%q): %w", id, err)
@@ -129,7 +131,10 @@ func resourceUserGroupAssociationDelete(d *schema.ResourceData, meta interface{}
 		UserIdsToRemove: aws.StringSlice([]string{d.Get("user_id").(string)}),
 	}
 
-	_, err := conn.ModifyUserGroup(input)
+	_, err := tfresource.RetryWhenAWSErrCodeEquals(10*time.Minute, func() (interface{}, error) {
+		return conn.ModifyUserGroup(input)
+	}, elasticache.ErrCodeInvalidUserGroupStateFault)
+
 	if err != nil && !tfawserr.ErrMessageContains(err, elasticache.ErrCodeInvalidParameterValueException, "not a member") {
 		return fmt.Errorf("deleting ElastiCache User Group Association (%q): %w", d.Id(), err)
 	}

--- a/internal/service/elasticache/user_group_association_test.go
+++ b/internal/service/elasticache/user_group_association_test.go
@@ -103,6 +103,35 @@ func TestAccElastiCacheUserGroupAssociation_disappears(t *testing.T) {
 	})
 }
 
+func TestAccElastiCacheUserGroupAssociation_multiple(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping long-running test in short mode")
+	}
+
+	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+	resourceName1 := "aws_elasticache_user_group_association.test1"
+	resourceName2 := "aws_elasticache_user_group_association.test2"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ErrorCheck:               acctest.ErrorCheck(t, elasticache.EndpointsID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckUserGroupAssociationDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccUserGroupAssociationConfig_preMultiple(rName),
+			},
+			{
+				Config: testAccUserGroupAssociationConfig_multiple(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckUserGroupAssociationExists(resourceName1),
+					testAccCheckUserGroupAssociationExists(resourceName2),
+				),
+			},
+		},
+	})
+}
+
 func testAccCheckUserGroupAssociationDestroy(s *terraform.State) error {
 	return testAccCheckUserGroupAssociationDestroyWithProvider(s, acctest.Provider)
 }
@@ -263,6 +292,44 @@ resource "aws_elasticache_user" "test3" {
 }
 
 resource "aws_elasticache_user_group_association" "test" {
+  user_group_id = aws_elasticache_user_group.test.user_group_id
+  user_id       = aws_elasticache_user.test3.user_id
+}
+`, rName))
+}
+
+func testAccUserGroupAssociationConfig_preMultiple(rName string) string {
+	return acctest.ConfigCompose(
+		testAccUserGroupAssociationBaseConfig(rName),
+		fmt.Sprintf(`
+resource "aws_elasticache_user" "test3" {
+  user_id       = "%[1]s-3"
+  user_name     = "username2"
+  access_string = "on ~app::* -@all +@read +@hash +@bitmap +@geo -setbit -bitfield -hset -hsetnx -hmset -hincrby -hincrbyfloat -hdel -bitop -geoadd -georadius -georadiusbymember"
+  engine        = "REDIS"
+  passwords     = ["password123456789"]
+}
+`, rName))
+}
+
+func testAccUserGroupAssociationConfig_multiple(rName string) string {
+	return acctest.ConfigCompose(
+		testAccUserGroupAssociationBaseConfig(rName),
+		fmt.Sprintf(`
+resource "aws_elasticache_user" "test3" {
+  user_id       = "%[1]s-3"
+  user_name     = "username2"
+  access_string = "on ~app::* -@all +@read +@hash +@bitmap +@geo -setbit -bitfield -hset -hsetnx -hmset -hincrby -hincrbyfloat -hdel -bitop -geoadd -georadius -georadiusbymember"
+  engine        = "REDIS"
+  passwords     = ["password123456789"]
+}
+
+resource "aws_elasticache_user_group_association" "test1" {
+  user_group_id = aws_elasticache_user_group.test.user_group_id
+  user_id       = aws_elasticache_user.test2.user_id
+}
+
+resource "aws_elasticache_user_group_association" "test2" {
   user_group_id = aws_elasticache_user_group.test.user_group_id
   user_id       = aws_elasticache_user.test3.user_id
 }

